chain/client/src/chunks_delay_tracker.rs

```diff
@@ -1,192 +1,129 @@
+    /// Timestamps of requesting missing chunks.
+    /// Chunks are usually requested when a client falls out-of-sync with the head of the chain.
+    chunk_requested: Option<Instant>,
 }
 
 impl ChunksDelayTracker {
-    // Blocks below this heights are not tracked.
-    fn lowest_height(head_height: BlockHeight) -> BlockHeight {
-        head_height.saturating_sub(CHUNKS_DELAY_TRACKER_HORIZON)
+    pub fn add_block_timestamp(&mut self, block_hash: &CryptoHash, timestamp: Instant) {
+        self.blocks_received.entry(*block_hash).or_insert(timestamp);
     }
 
-    fn remove_old_entries(&mut self, head_height: BlockHeight) {
-        let heights_to_drop: Vec<BlockHeight> = self
-            .heights
-            .range(..ChunksDelayTracker::lowest_height(head_height))
-            .map(|(&k, _v)| k)
-            .collect();
-        for h in heights_to_drop {
-            self.heights.remove(&h);
-        }
-    }
-
-    // Computes the delay between receiving a block and receiving the latest of its chunks.
-    // Updates the metric to the max delay across the most recently processed blocks.
-    fn get_max_delay(&mut self, head_height: BlockHeight) -> Duration {
-        let mut max_delay = Duration::ZERO;
-        for (_, v) in self.heights.range(..=head_height) {
-            if let Some(block_received) = v.block_received {
-                if let Some(latest_chunk) = v.chunks_received.values().max() {
-                    if latest_chunk > &block_received {
-                        let delay = latest_chunk.duration_since(block_received);
-                        max_delay = max(max_delay, delay);
-                    }
-                }
+    pub fn add_chunk_timestamp(&mut self, chunk_hash: &ChunkHash, timestamp: Instant) {
+        match self.chunks_in_progress.get_mut(chunk_hash) {
+            Some(entry) => {
+                entry.chunk_received.get_or_insert(timestamp);
+            }
+            None => {
+                self.chunks_in_progress.put(
+                    chunk_hash.clone(),
+                    ChunkInProgress { chunk_received: Some(timestamp), chunk_requested: None },
+                );
             }
         }
-        max_delay
-    }
-
-    fn update_chunks_metric(&mut self, head_height: BlockHeight) {
-        metrics::CHUNKS_RECEIVING_DELAY_US.set(self.get_max_delay(head_height).as_micros() as i64);
     }
 
-    // Computes the difference between the latest block we are aware of and the current head.
-    fn get_blocks_ahead(&mut self, head_height: BlockHeight) -> u64 {
-        if let Some((&latest, _)) = self.heights.iter().next_back() {
-            latest.saturating_sub(head_height)
-        } else {
-            0
+    pub fn requested_chunk(&mut self, chunk_hash: &ChunkHash, timestamp: Instant) {
+        match self.chunks_in_progress.get_mut(chunk_hash) {
+            Some(entry) => {
+                entry.chunk_requested.get_or_insert(timestamp);
+            }
+            None => {
```

Comment:
Similarly, you can get rid of the "Some" case here

Reply:
Done